src/Microsoft.Android.Sdk.ILLink/Microsoft.Android.Sdk.ILLink.csproj

```diff
@@ -41,9 +41,6 @@
 
     <!--Other Xamarin.Android / Java.Interop files-->
     <Compile Include="..\..\external\Java.Interop\src\Java.Interop.Tools.Cecil\Java.Interop.Tools.Cecil\CustomAttributeProviderRocks.cs" Link="Java.Interop\CustomAttributeProviderRocks.cs" />
-    <Compile Include="..\..\external\Java.Interop\src\Java.Interop.Tools.Cecil\Java.Interop.Tools.Cecil\MethodDefinitionRocks.cs" Link="Java.Interop\MethodDefinitionRocks.cs" />
-    <Compile Include="..\..\external\Java.Interop\src\Java.Interop.Tools.Cecil\Java.Interop.Tools.Cecil\TypeDefinitionCache.cs" Link="Java.Interop\TypeDefinitionCache.cs" />
```

Comment:
Why do we want to remove `TypeDefinitionCache`?  It's inclusion had a nice performance win; see also xamarin/java.interop@b81cfbb9ab8efa647a3bfcc2b2b97a5f1b1fa71e.

Now, perhaps it's "not needed" because `LinkContext` does the same thing…but does it?

Also, where *is* `LinkContext` defined?  The [`LinkContext.cs` in mono/linker](https://github.com/mono/linker/blob/a621b9a630950026f5508c8089a59c0e825bd6e0/src/linker/Linker/LinkContext.cs) doesn't contain an `AreParametersCompatibleWith()` method, e.g. used here: https://github.com/xamarin/xamarin-android/pull/5748/files#diff-973bf7b778beba59537f406bfc5abcacba0e5d76aac82068a2d15908e8fd5db8R22

Reply:
See https://github.com/xamarin/xamarin-android/pull/5748#discussion_r606634212 for context - I unresolved it in case you want to continue the discussion there.

The idea was to move to a similar cache that's built-in to the linker. `AreParametersCompatibleWith` is defined in `LinkContextExtensions.cs` (copied from java.interop's [MethodDefinitionRocks](https://github.com/xamarin/java.interop/blob/main/src/Java.Interop.Tools.Cecil/Java.Interop.Tools.Cecil/MethodDefinitionRocks.cs#L54) and adapted to use the built-in cache).

Reply:
Also, re-asking:

Where *is* `LinkContext` defined?  The [`LinkContext.cs` in mono/linker](https://github.com/mono/linker/blob/a621b9a630950026f5508c8089a59c0e825bd6e0/src/linker/Linker/LinkContext.cs) doesn't contain an `AreParametersCompatibleWith()` method, e.g. used here: https://github.com/xamarin/xamarin-android/pull/5748/files#diff-973bf7b778beba59537f406bfc5abcacba0e5d76aac82068a2d15908e8fd5db8R22

Furthermore, I can't find `AreParametersCompatibleWith()` on *any* mono/linker branch:

```
$ git grep AreParametersCompatibleWith $(git rev-list --all)
# no matches
```

This is on a just-fetched repo, with HEAD at mono/linker@e069cd8d.

How does this PR even *build*?

/me checks PR status, realizes that it hasn't even been built as of the most recent commits to the branch. :thinking:

Reply:
> Furthermore, I can't find `AreParametersCompatibleWith()` on any mono/linker branch:

:rofl: it's the extension method in this PR!

https://github.com/xamarin/xamarin-android/pull/5748/files#diff-973bf7b778beba59537f406bfc5abcacba0e5d76aac82068a2d15908e8fd5db8R30

/me hangs head in shame.

Reply:
@sbomer: I wrote this up: https://github.com/xamarin/java.interop/pull/842

The idea being that if you also update `LinkContext` to likewise implement `IMetadataResolver`, most/all of `LinkContextExtensions.cs` won't be needed, removing "one" of the algorithmic duplications.

Reply:
@jonpryor Thanks, I like this idea a lot. To make sure I understand how this fits together, is this an accurate summary?
- java.interop will use `TypeDefinitionCache` as the `IMetadataResolver`
- the xamarin-android linker steps will use `LinkContext` as the `IMetadataResolver`
  - which uses the linker's built-in caching instead of `TypeDefinitionCache`
  - `LinkContext` doesn't take any extra ctor arguments

Reply:
> it's the extension method in this PR!

😆 Sorry, I should've linked to it in my answer!

> realizes that it hasn't even been built as of the most recent commits to the branch.

I just have to remember to ping @jonathanpeppers each time I want to trigger ci

Reply:
@sbomer wrote:

> To make sure I understand how this fits together…

There are two similar-yet-distinct answers here.

Option 1, proposed by @jonathanpeppers [on Discord](https://discord.com/channels/732297728826277939/732297837953679412/844308968243593216):

> Could we just subclass `TypeDefinitionCache` and override `Resolve()` so it uses the `LinkContext`?

```csharp
partial class LinkContextTypeDefinitionCache : TypeDefinitionCache {
    LinkContext context;

    public LinkContextTypeDefinitionCache (LinkContext context)
    {
        this.context = context;
    }

    public override TypeDefinition Resolve (TypeReference typeReference)
    {
         return context.ResolveTypeDefinition (typeReference);
    }
}
```

This would require *no* changes to Java.Interop.

Option 2 is to merge https://github.com/xamarin/java.interop/pull/842 and then either:

1. Update `LinkContext` to implement `IMetadataResolver`, or
2. Add a new `LinkContextMetadataResolver` type, along the lines of the above `LinkContextTypeDefinitionCache` type, which implements `IMetadataResolver` in terms of `LinkContext`.

Either/both of these would remove the need for `LinkContextExtensions.cs`, and should allow using the existing `GetBaseDefinition()` extension method from Java.Interop without any changes, and would remove the need for `LinkContextExtensions.cs`.  (I think.)

Is there a preference between (1) or (2)?  Of course, *Both* is a viable answer: (1) could be used while waiting for (2) to be completed, and if we choose the same name -- `LinkContextMetadataResolver`? -- then future "integration" changes are reduced in scope.

Reply:
@sbomer: related question: are you actively on Discord?  Your nick appears to be there, but I don't see any messages from you in recent history.  Or, are you a *different* `@sbomer` than is on our Discord? https://discord.gg/95Pww5tQ

Reply:
Yup that's me! I haven't been active on it, but I will start checking it more frequently.

Reply:
Thanks for clarifying. I like the combination of the options. I implemented 1 (with the name `LinkContextMetadataResolver`) here  for now. I also have a PR out for option 2.1, so we can see what @marek-safar thinks of implementing `IMetadataResolver` directly on `LinkContext`.